### PR TITLE
Ct 352 - Add Service Feedback

### PR DESCRIPTION
**Pull request by aliuk2012**
https://dsdmoj.atlassian.net/browse/CT-352

- Adds Feedback model
- Add JS to to AJAX to submit the feedback form without submitting the whole page
- Setup Poltergeist for AJAX/JS Testing
- Allow shared DB Connections for tests. (Please check https://github.com/ministryofjustice/correspondence_tool_staff/pull/100/commits/658f19cb892ab62d9bbd9a2c5d225c16ce809e46#diff-f2369014861328398fb9a81dd6426247 as mentioned in the answer to this question http://stackoverflow.com/questions/18623661/why-is-capybara-discarding-my-session-after-one-event)

### CODE REVIEW

File: spec/features/00_feedback/submit_feedback_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    expect(home_page).to have_service_feedback
+    expect(home_page.service_feedback).to have_feedback_form
+    expect(home_page.service_feedback).to have_send_button
+
+    home_page.service_feedback.feedback_textarea.set "Very good service"
+    home_page.service_feedback.send_button.click
+
+    home_page.service_feedback.wait_for_success_notice
+    expect(home_page.service_feedback).to have_success_notice
+  end
```
Comment (EdwardAndress) [inline spec/features/00_feedback/submit_feedback_spec.rb]: What happens to the feedback box when feedback is submitted?  Is the text cleared?

Comment (EdwardAndress) [inline spec/features/00_feedback/submit_feedback_spec.rb]: Should we include some expectations which check that the feedback was saved?

File: spec/controllers/feedback_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      post :create, params: params, xhr: true
+    }
+
+    it "return 200 if trying to submit new feedback using AJAX" do
+      expect(response.code).to eq "200"
+    end
+
+    it 'makes a DB entry' do
+      expect(Feedback.first.comment).to eq "This is my feedback"
+    end
```
Comment (EdwardAndress) [inline spec/controllers/feedback_controller_spec.rb]: Should we check the other attributes too?

File: spec/support/pages/feedback_section.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class ServiceFeedback < SitePrism::Section
+  element :feedback_form, '#new_feedback'
+  element :feedback_textarea, '#new_feedback textarea'
+
+  element :send_button, '.button-secondary'
```
Comment (EdwardAndress) [inline spec/support/pages/feedback_section.rb]: As an aside, maybe the upload documents button should be a secondary button, because it doesn't cause a state transition...

File: app/controllers/feedback_controller.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+  private
+
+  def create_feedback_params
+    params.require(:feedback).permit(
+        :comment,
+    ).merge({ email: current_user.email,
+              user_agent: request.user_agent,
+              referer: request.referer
+            })
```
Comment (EdwardAndress) [inline app/controllers/feedback_controller.rb]: Do we test anywhere that the 'referer' and  user_agent are set correctly?  Are there any circumstances in which these might be nil?  Will the application handle that?

Comment (misaka) [inline app/controllers/feedback_controller.rb]: I'm not sure that we can test that these are set correctly ... it's in-built Rails functionality that relies on headers in the HTTP request. We could try to catch the edge-case where they're set to `nil` to make sure it doesn't break the user journey, but then, if they're `nil` then there's probably something abnormal going on anyway (the user is actually a script, for example).

File: app/controllers/feedback_controller.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
+class FeedbackController < ApplicationController
+
+  def create
+    @feedback = Feedback.new(create_feedback_params)
+
+    respond_to do |format|
+      if @feedback.save
```
Comment (misaka) [inline app/controllers/feedback_controller.rb]: Why an `if` statement here if the behaviour doesn't change?

File: app/assets/javascripts/modules/Feedback.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+moj.Modules.Feedback = {
+  init: function(){
```
Comment (misaka) [inline app/assets/javascripts/modules/Feedback.js]: How about putting this logic into the `create.js.erb` below to keep the response logic the same?

File: app/views/feedback/_alerts.html.slim [hunk 7]
```diff
@@ -0,0 +1,12 @@
+- if feedback.errors.any?
+  .alert-red
+    | The following error(s) occurred:
```
Comment (misaka) [inline app/views/feedback/_alerts.html.slim]: This line looks like a candidate for I18n.

File: spec/controllers/feedback_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
+require 'rails_helper'
+
+RSpec.describe FeedbackController, type: :controller do
+
+  let(:params)do
+    { feedback: {
+        comment: "This is my feedback",
+        user_agent: 'firefox'
+    }
+    }
```
Comment (misaka) [inline spec/controllers/feedback_controller_spec.rb]: I know this is getting picky, but having two braces like this makes me immediately think that there's been a cut'n'paste error. I know we're not being terribly consistent with our code style here, which is ok, but could we at least avoid false-alarms like this? I would probably bust it out so that `feedback: {` is on a separate line, and then the first brace should line-up with that.

File: spec/controllers/feedback_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  end
+
+  context "as an anonymous user" do
+
+    describe '#create' do
```
Comment (misaka) [inline spec/controllers/feedback_controller_spec.rb]: Could we move `context` and `describe` around so that instead of having:

```
context 'as an anonymous user' do
  describe '#create' do
```

```
describe '#create' do
  context 'as an anonymous user' do
```

So we can group together tests around *what* we're testing, instead of the various contexts we'll have?

File: app/controllers/feedback_controller.rb (hunk 4 above)
Comment (aliuk2012) [inline app/controllers/feedback_controller.rb]: refer and user_agents aren't required attributes so if they are nil, then so be it they will just be recorded as nil. It just saves us having to contact the user to ask them things like "What page were you on and what browser were you using"

File: spec/models/feedback_spec.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+RSpec.describe Feedback, type: :model do
+
+  let(:feedback) { build :feedback}
+
+  describe 'has a factory' do
```
Comment (misaka) [inline spec/models/feedback_spec.rb]: Is this testing that FactoryGirl works? Isn't this overkill?

File: spec/features/00_feedback/submit_feedback_spec.rb (hunk 1 above)
Comment (aliuk2012) [inline spec/features/00_feedback/submit_feedback_spec.rb]: 👍

File: spec/controllers/feedback_controller_spec.rb (hunk 2 above)
Comment (aliuk2012) [inline spec/controllers/feedback_controller_spec.rb]: I've added emails as we need to know who it has come from.

File: spec/support/pages/feedback_section.rb (hunk 3 above)
Comment (aliuk2012) [inline spec/support/pages/feedback_section.rb]: Trick question, Upload button is primary if they have not uploaded any previous responses but once they have I think that switches to a secondary action as the next stage is for them to mark the case as responded

File: spec/controllers/feedback_controller_spec.rb (hunk 8 above)
Comment (aliuk2012) [inline spec/controllers/feedback_controller_spec.rb]: 👍

File: spec/models/feedback_spec.rb (hunk 10 above)
Comment (aliuk2012) [inline spec/models/feedback_spec.rb]: I don't think this is overkill. All the other models spec have similar syntax but I think its good to make sure that models have a valid factory to use and test with

**Comment by aliuk2012** [comment]
@EdwardAndress @misaka I think I've addressed all your comments. Could you review?
